### PR TITLE
ARROW-3377: [Gandiva][C++] Replace If statement with bit operations for bitmap

### PR DESCRIPTION
**Pull request by Praveen2112**
This patch replaces the existing If statement with a bit operations for optimal performance

### CODE REVIEW

File: cpp/src/gandiva/precompiled/bitmap.cc [hunk 1]
```diff
@@ -38,11 +38,7 @@ FORCE_INLINE
 void bitMapSetBit(unsigned char* bmap, int position, bool value) {
   int byteIdx = POS_TO_BYTE_INDEX(position);
   int bitIdx = POS_TO_BIT_INDEX(position);
-  if (value) {
-    bmap[byteIdx] |= (1 << bitIdx);
-  } else {
-    bmap[byteIdx] &= ~(1 << bitIdx);
-  }
+  bmap[byteIdx] ^= (-value ^ bmap[byteIdx]) & (1UL << bitIdx);
```
Comment (wesm) [inline cpp/src/gandiva/precompiled/bitmap.cc]: There is some redundancy here with arrow/util/bit-util.h. We should explore merging these together

File: cpp/src/gandiva/precompiled/bitmap.cc [hunk 2]
```diff
@@ -38,11 +38,7 @@ FORCE_INLINE
 void bitMapSetBit(unsigned char* bmap, int position, bool value) {
```
Comment (wesm) [inline cpp/src/gandiva/precompiled/bitmap.cc]: NB the function name and variable names in this function don't conform to the Arrow style guide

File: cpp/src/gandiva/precompiled/bitmap.cc (hunk 1 above)
Comment (Praveen2112) [reply cpp/src/gandiva/precompiled/bitmap.cc]: Yeah but AFAIK this code comes under the pre-compiled part of gandiva where we create a LLVM-IR code for the above bitmap operation and use them in the runtime. @pravindra Your insights ? Can we merge both of them ?

File: cpp/src/gandiva/precompiled/bitmap.cc (hunk 2 above)
Comment (Praveen2112) [reply cpp/src/gandiva/precompiled/bitmap.cc]: Is there is any documents that explains the codestyle that we use in arrow ?

File: cpp/src/gandiva/precompiled/bitmap.cc (hunk 1 above)
Comment (pravindra) [reply cpp/src/gandiva/precompiled/bitmap.cc]: agree with @wesm. but, it'll require some changes to use this file in the IR generation (eg. we don't want the functions to be inlined at compile time, but later in the optimization pass).

@wesm, Can we use a seperate jira for merging the files ?

Comment (wesm) [reply cpp/src/gandiva/precompiled/bitmap.cc]: Yes, separate JIRA is fine

**Comment by pitrou** [comment]
I think you should post benchmark numbers to validate that this does speed things up significantly. The compiler might be smarter than you are expecting.

**Comment by wesm** [comment]
Indeed. Once we get the repo caught up with the Gandiva patches we should work together to be more rigorous about microbenchmarking so we can iterate on performance in a safe and monotonic fashion

**Pull request closed without merging**